### DISCUSSION TITLE
Support for stereo swap chains

### DISCUSSION THREAD

**Discussion by hals1010** [DiligentGraphics/DiligentEngine#159] [Ideas]
I'm using stereo swap chains. Unfortunately, Diligent's SwapChainDesc struct has no member to request that feature from
pFactory->CreateSwapChainXXX().

Would it be possible to add support for this?

From the DirectX 11 perspective, a stereo swap chain is created like this:

```
DXGI_SWAP_CHAIN_DESC1    Desc = { 0, };
...
Desc.Stereo                = TRUE;
pDXGIFactory2->CreateSwapChainForHwnd(pDevice, hWnd, &Desc, nullptr, nullptr, &pSwapChainStereo);

```

You'll end up with a 2D texture array for the render target (0: left eye, 1: right eye):

```
CD3D11_RENDER_TARGET_VIEW_DESC ViewDescLeft(D3D11_RTV_DIMENSION_TEXTURE2DARRAY, eFormat, 0, 0, 1);
CD3D11_RENDER_TARGET_VIEW_DESC ViewDescRight(D3D11_RTV_DIMENSION_TEXTURE2DARRAY, eFormat, 0, 1, 1);

ID3D11RenderTargetView*    pRenderTargetViewLeft;
ID3D11RenderTargetView*    pRenderTargetViewRight;

pDevice->CreateRenderTargetView(pColorBuffer, &ViewDescLeft, &pRenderTargetViewLeft);
pDevice->CreateRenderTargetView(pColorBuffer, &ViewDescRight, &pRenderTargetViewRight);

```

I suppose one has to extend the GetDefaultView() method somehow:

```
auto    pRTV = pTexRenderTarget->GetDefaultView(TEXTURE_VIEW_RENDER_TARGET);    // current
auto    pRTVStereo = pTexRenderTarget->GetDefaultView(TEXTURE_VIEW_RENDER_TARGET, nEyeIndex);    // new

```

Thanks,
Jörn

**Comment by TheMostDiligent** [comment]
How is the stereo swap chain connected to the HMD (I suppose you are using it for VR?)

**Comment by hals1010** [comment]
It's not VR, but a 3D "window" in a 2D UI. We use it for some sort of CAD (120 Hz monitor + shutter glasses, for instance).

**Reply by TheMostDiligent** [reply]
Why do you need a stereo swap chain? Why can't you use two swap chains instead? Or one swap chain with twice the window size?

**Reply by hals1010** [reply]
"Stereo" does the magic of synchronizing the hardware and the left eye / right eye backbuffers when calling Present().

**Reply by TheMostDiligent** [reply]
This seems to be too API/use-case specific to add this to the common API at this point. What I suggest is to create the native swap D3D11 chain, retrieve back buffers from it and then attach Diligent textures to them. This way you will be able to use Diligent API with the stereo D3D11 swap chain.

**Reply by hals1010** [reply]
Thanks for the suggestion, will give it a try.

But may I add that "stereo" is not a DirectX-11-only feature? I remember an OpenGL feature called "quad buffering", incl. enums like
```
GL_FRONT_LEFT
GL_FRONT_RIGHT
GL_BACK_LEFT
GL_BACK_RIGHT

```
I'm pretty sure Vulkan and DirectX 12 have similar approaches (haven't tried yet)
